tensorflow/python/keras/layers/preprocessing/preprocessing_stage_functional.py

```diff
@@ -0,0 +1,141 @@
+                                   functional.Functional):
+  """A functional preprocessing stage.
+
+  This preprocessing stage wraps a graph of preprocessing layers into a
+  Functional-like object that enables you to `adapt()` the whole graph via
+  a single `adapt()` call on the preprocessing stage.
+
+  Arguments:
+    inputs: List of input tensors (must be created via `tf.keras.Input()`).
+    outputs: List of outputs tensors.
+    name: String, optional. Name of the model.
+    trainable: Boolean, whether the model's variables should be trainable.
+  """
+
+  def adapt(self, data, reset_state=True):
+    """Adapt the state of the layers of the preprocessing stage to the data.
+
+    Arguments:
+      data: A batched Dataset object, or a NumPy array, or an EagerTensor.
+        Data to be iterated over to adapt the state of the layers in this
+        preprocessing stage.
+      reset_state: Whether this call to `adapt` should reset the state of
+        the layers in this preprocessing stage.
+    """
+    if not isinstance(data, dataset_ops.Dataset):
+      data = self._flatten_to_reference_inputs(data)
+      if any([not isinstance(datum, (np.ndarray, ops.EagerTensor))
+              for datum in data]):
+        raise ValueError(
+            '`adapt()` requires a batched Dataset, a list of EagerTensors '
+            'or Numpy arrays as input, got {}'.format(type(data)))
+
+    if isinstance(data, dataset_ops.Dataset):
+      # Validate the datasets to try and ensure we haven't been passed one with
+      # infinite size. That would cause an infinite loop here.
+      if tf_utils.dataset_is_infinite(data):
+        raise ValueError(
+            'The dataset passed to `adapt()` has an infinite number of '
+            'elements. Please use dataset.take(...) to make the number '
+            'of elements finite.')
+      # Unzip dataset object to a list of single input dataset.
+      data = _unzip_dataset(data)
+
+    # Dictionary mapping reference tensors to data
+    data_dict = {}
+    tensor_usage_count = self._tensor_usage_count
+    for x, y in zip(self.inputs, data):
+      if isinstance(y, ops.EagerTensor):
+        y = self._conform_to_reference_input(y, ref_input=x)
+      x_id = str(id(x))
+      data_dict[x_id] = [y] * tensor_usage_count[x_id]
+
+    nodes_by_depth = self._nodes_by_depth
+    depth_keys = sorted(nodes_by_depth.keys(), reverse=True)
+    for depth in depth_keys:
+      for node in nodes_by_depth[depth]:
+        # Input node
+        if node.is_input:
+          continue
+
+        # Node with input not computed yet
+        if any(t_id not in data_dict for t_id in node.flat_input_ids):
+          continue
+
+        args, kwargs = node.map_arguments(data_dict)
+
+        if hasattr(node.layer, 'adapt'):
+          args_flatten = nest.flatten(args)
+          if all(isinstance(arg, dataset_ops.Dataset) for arg in args_flatten):
+            args = dataset_ops.Dataset.zip(tuple(args_flatten))
```

Comment:
`zip` is not necessarily the right thing to do, the dataset should be conformed to the input format expected by the layer.

Reply:
Before `zip`, `args` has the structure required by the layer, with each element being a separate dataset object. So now I use `zip` directly on `args` without flatten it. What do you think?
I tried `zip` on situations for layers that takes single input (`x`), list of inputs (`[x, y]`), nested inputs (`[[x, y], z]`) and dict of inputs (`{'x':x, 'y':y}`). Seems `zip` preserves input format correctly.